Add docs for `local_testing_mode` (#48567)
ray-project/ray: doc/source/serve/advanced-guides/dev-workflow.md

(ServeController pid=9865) INFO 2022-08-11 11:47:19,929 controller 9865 deployment_state.py:1257 - Removing 1 replicas from deployment 'HelloDeployment'.
```

Note that rerunning `serve run` redeploys all deployments. To prevent redeploying the deployments whose code hasn't changed, you can use `serve deploy`; see the [Production Guide](serve-in-production) for details.

> ### Local Testing Mode

Reviewer: can you please add a markdown `NOTE` (see other pages for example syntax) that indicates this feature is experimental

> Note that rerunning `serve run` will redeploy all deployments. To prevent redeploying those deployments whose code hasn't changed, you can use `serve deploy`; see the [Production Guide](serve-in-production) for details.

Reviewer 2: ```suggestion
Note that rerunning `serve run` redeploys all deployments. To prevent redeploying the deployments whose code hasn't changed, you can use `serve deploy`; see the [Production Guide](serve-in-production) for details.
```

### Local Testing Mode

> Ray Serve now supports a local testing mode that allows you to run your deployments locally in a single process. This mode is useful for unit testing and debugging your application logic without the overhead of a full Ray cluster. To enable this mode, use the `_local_testing_mode` flag in the `serve.run` function:

Reviewer 2: ```suggestion
Ray Serve supports a local testing mode that allows you to run your deployments locally in a single process. This mode is useful for unit testing and debugging your application logic without the overhead of a full Ray cluster. To enable this mode, use the `_local_testing_mode` flag in the `serve.run` function:
```

```python

> serve.run(app, _local_testing_mode=True)

Reviewer 3: Kinda nit, but we should move this code into `local_dev.py` so it's tested correctly. You can follow this as an example to link the code back https://github.com/ray-project/ray/blob/17579cba9b793c2eccd2937d7a58e26ce22646db/doc/source/serve/advanced-guides/dev-workflow.md?plain=1#L31-L35

```

Alternatively, you can set the environment variable `RAY_SERVE_FORCE_LOCAL_TESTING_MODE=1`.

> This mode runs each deployment in a background thread and supports most of the same features as running on a full Ray cluster. Note that some features, such as converting `DeploymentResponses` to `ObjectRefs`, are not supported in local testing mode. If you encounter limitations, consider filing a feature request on GitHub.

Reviewer 2: ```suggestion
This mode runs each deployment in a background thread and supports most of the same features as running on a full Ray cluster. Note that some features, such as converting `DeploymentResponses` to `ObjectRefs`, aren't supported in local testing mode. If you encounter limitations, consider filing a feature request on GitHub.
```

## Testing on a remote cluster

> To test on a remote cluster, you'll use `serve run` again, but this time you'll pass in an `--address` argument to specify the address of the Ray cluster to connect to.  For remote clusters, this address has the form `ray://<head-node-ip-address>:10001`; see [Ray Client](ray-client-ref) for more information.

Reviewer 2: ```suggestion
To test on a remote cluster, use `serve run` again, but this time, pass in an `--address` argument to specify the address of the Ray cluster to connect to.  For remote clusters, this address has the form `ray://<head-node-ip-address>:10001`; see [Ray Client](ray-client-ref) for more information.
```
